### DISCUSSION TITLE
How to rewrite express Router to api-routes?

### DISCUSSION THREAD

**Discussion by popuguytheparrot** [vercel/next.js#16438] [Help]
Hi. I want to remove custom server to api-routes and dont understand how do it. 

I have url `api/auth/login` and in custom server handle it like this

```ts
router.use('/api/auth/login', authRouter);
```
### authRouter
```ts
authRouter.get('/google', getSocialAuthHandler('google', { scope: ['profile', 'email'] }));
authRouter.get('/google/callback', getSocialAuthCallbackHandlers('google'));

authRouter.get('/fb', getSocialAuthHandler('facebook', { scope: ['email'] }));
authRouter.get('/fb/callback', getSocialAuthCallbackHandlers('facebook'));
```

I try create folder `api/auth/login.ts` and move authRouter to .ts file with default export
But it doesnt work
```ts
export default authRouter
```

So, how to handle my case right with api-routes?

**Comment by jamesmosier** [comment]
Next API routes are [defined using the `pages/api` structure](https://nextjs.org/docs/api-routes/introduction), so you must create a new file for each route:

```
pages/api/auth/login/
  google/
    index.js
    callback.js
  fb/
    index.js
    callback.js  
```

**Reply by popuguytheparrot** [reply]
I dont want copy paste code and create excessive files. Maybe there are other solutions without custom server and copy paste ?

**Reply by jamesmosier** [reply]
You can create helper functions outside of the `pages/api/*` folder that can be reused or create [middleware](https://nextjs.org/docs/api-routes/api-middlewares) like [this example](https://github.com/vercel/next.js/tree/canary/examples/api-routes-middleware)

`lib/getSocialAuthCallbackHandlers.js`

```js
const getSocialAuthCallbackHandlers = (handler, socialMediaType) => (req, res) => {
  // some logic
  return handler(req, res)
}
export default getSocialAuthCallbackHandlers
```

`pages/api/auth/login/google/callback.js`

```js
function GoogleCallback(req, res) {
 // your api route logic
}

export default getSocialAuthCallbackHandlers(GoogleCallback, 'google')
```

**Reply by popuguytheparrot** [reply]
Thanks for the suggestion, but that doesn't solve the creation of extra files.

**Reply by jamesmosier** [reply]
What Express does is not able to be reproduced exactly with Next.

You could also use [dynamic API routes](https://nextjs.org/docs/api-routes/dynamic-api-routes) such as `pages/api/auth/login/[type]/index.js` and `pages/api/auth/login/[type]/callback.js` and perform conditional logic depending on what parameter is passed in.

**Reply by popuguytheparrot** [reply]
i will try some more solutions and as a last resort return to the ones you suggested
